paddle/fluid/operators/set_value_op_npu.cc

```diff
@@ -0,0 +1,392 @@
+#include "paddle/fluid/operators/utils.h"
+#include "paddle/fluid/platform/enforce.h"
+
+namespace paddle {
+namespace operators {
+template <typename DeviceContext, typename T>
+class SetValueNPUKernel : public framework::OpKernel<T> {
+ private:
+  using vec64 = std::vector<int64_t>;
+  using vec_vec64 = std::vector<std::vector<int64_t>>;
+  void GetNPUStartEndSteps(const vec64& start, const vec64& end,
+                           const vec64& steps, const vec64& axes,
+                           const framework::DDim in_dim,
+                           vec_vec64& output) const {
+    int rank = in_dim.size();
+    for (int i = 0; i < rank; ++i) {
+      int axis_size = in_dim[i];
+      auto iter = find(axes.begin(), axes.end(), i);
+      if (iter != axes.end()) {
+        int idx = iter - axes.begin();
+        output[0].push_back(start[idx]);  // set as the same as raw input
+        output[1].push_back(end[idx]);
+        output[2].push_back(steps[idx]);
+      } else {
+        output[0].push_back(0);          // begin 0
+        output[1].push_back(axis_size);  // end = last one
+        output[2].push_back(1);          // step = 1
+      }
+    }
+  }
+
+  inline std::vector<int> MininumPadNumberMakeSureLastDimGT8(
+      const vec_vec64& npu_slice) const {
+    int min_value = min_last_dim_value;
+    int rank = npu_slice[0].size();
+    int last_dim_start = npu_slice[0][rank - 1];
+    int last_dim_end = npu_slice[1][rank - 1];
+    int last_dim_step = npu_slice[2][rank - 1];
+    int min_end = last_dim_start + last_dim_step * min_value;
+    int raw_last_dim_len = (last_dim_end - last_dim_start) / last_dim_step;
+    return std::vector<int>(
+        {std::max(0, min_end - last_dim_end), min_value - raw_last_dim_len});
+  }
+
+  inline void TileTensor(const framework::ExecutionContext* ctx,
+                         const Tensor* input, Tensor* output) const {
+    VLOG(4) << "start to tile tensor function, which calls the npu operator "
+               "TileWithAxis";
+    // UNSQUEEZE last dim + TILE last dim * min_last_dim_value
+    Tensor reshape_tensor;
+    auto reshape_dims = framework::vectorize<int>(input->dims());
+    reshape_dims.push_back(1);
+    reshape_tensor.ShareDataWith(*input);
+    reshape_tensor.Resize(framework::make_ddim(reshape_dims));
+
+    auto output_dims = framework::vectorize<int>(input->dims());
+    output_dims.push_back(min_last_dim_value);
+    output->mutable_data<T>(framework::make_ddim(output_dims), ctx->GetPlace());
+
+    framework::NPUAttributeMap attr;
+    attr["axis"] = static_cast<int>(reshape_dims.size() - 1);
+    attr["tiles"] = min_last_dim_value;
+    auto stream =
+        ctx->template device_context<paddle::platform::NPUDeviceContext>()
+            .stream();
+    NpuOpRunner("TileWithAxis", {reshape_tensor}, {*output}, attr).Run(stream);
+  }
+
+  inline void BroadcastToD(const framework::ExecutionContext* ctx,
+                           const Tensor* input,
```

Comment:
Better use `const Tensor &` because reference is always valid and we don't need to add `if` statement. But if you use `const Tensor*` ，you should ensure that it's not a `nullptr`